com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -336,60 +409,88 @@ public void Store(ulong networkObjectId, int behaviourIndex, int variableIndex,
             k.NetworkObjectId = networkObjectId;
             k.BehaviourIndex = (ushort)behaviourIndex;
             k.VariableIndex = (ushort)variableIndex;
+            k.TickWritten = m_NetworkManager.NetworkTickSystem.GetTick();
 
             int pos = m_Snapshot.Find(k);
             if (pos == Entry.NotFound)
             {
-                pos = m_Snapshot.AddEntry(networkObjectId, behaviourIndex, variableIndex);
+                pos = m_Snapshot.AddEntry(k);
             }
 
-            // write var into buffer, possibly adjusting entry's position and length
+            WriteVariableToSnapshot(m_Snapshot, networkVariable, pos);
+        }
+
+        private void WriteVariableToSnapshot(Snapshot snapshot, INetworkVariable networkVariable, int index)
+        {
+            // write var into buffer, possibly adjusting entry's position and Length
             using (var varBuffer = PooledNetworkBuffer.Get())
             {
                 networkVariable.WriteDelta(varBuffer);
-                if (varBuffer.Length > m_Snapshot.Entries[pos].Length)
+                if (varBuffer.Length > snapshot.Entries[index].Length)
                 {
                     // allocate this Entry's buffer
-                    m_Snapshot.AllocateEntry(ref m_Snapshot.Entries[pos], varBuffer.Length);
+                    snapshot.AllocateEntry(ref snapshot.Entries[index], index, (int)varBuffer.Length);
                 }
 
-                m_Snapshot.Entries[pos].TickWritten = m_NetworkManager.NetworkTickSystem.GetTick();
                 // Copy the serialized NetworkVariable into our buffer
-                Buffer.BlockCopy(varBuffer.GetBuffer(), 0, m_Snapshot.Buffer, m_Snapshot.Entries[pos].Position, (int)varBuffer.Length);
+                Buffer.BlockCopy(varBuffer.GetBuffer(), 0, snapshot.Buffer, snapshot.Entries[index].Position, (int)varBuffer.Length);
             }
         }
 
+
         /// <summary>
         /// Entry point when a Snapshot is received
         /// This is where we read and store the received snapshot
         /// </summary>
+        /// <param name="clientId">
         /// <param name="snapshotStream">The stream to read from</param>
-        public void ReadSnapshot(Stream snapshotStream)
+        public void ReadSnapshot(ulong clientId, Stream snapshotStream)
         {
+            ushort snapshotTick = default;
+
             using (var reader = PooledNetworkReader.Get(snapshotStream))
             {
-                ReadIndex(reader);
-                ReadBuffer(reader, snapshotStream);
+                snapshotTick = reader.ReadUInt16();
+
+                if (!m_ClientReceivedSnapshot.ContainsKey(clientId))
+                {
+                    m_ClientReceivedSnapshot[clientId] = new Snapshot(m_NetworkManager, false);
```

Comment:
Shouldn't there be a corresponding removal from `m_ClientReceivedSnapshot` when a client disconnects?

Reply:
This is another part that goes away in the `spawn` branch. Since there's a single snapshot, we don't allocate per client.